File: gfg/2024_8_august/c++/08_sum_tree.cpp
// TIME COMPEXITY
// O(n)

// SPACE COMPEXITY
// O(n)

// CODE

class Solution {
  public:
    pair<bool,int> solve(Node * root){
        if(root == NULL){
            pair<bool,int> p = make_pair(true,0);
            return p;
        }
        if(root->left == NULL && root->right == NULL){
            pair<bool,int> p = make_pair(true,root->data);
            return p;
        }
        
        pair<bool,int> left = solve(root->left);
        pair<bool,int> right = solve(root->right);
        
        bool leftAns = left.first;
        bool rightAns = right.first;
        
        bool condition = (root->data == left.second + right.second);
        
        pair<bool,int> ans;
        if(leftAns && rightAns && condition){
            ans.first = true;
            ans.second = 2 * root->data;
        }
        else{
            ans.first = 0;
        }
        
        return ans;
    }
    bool isSumTree(Node* root) {
        return solve(root).first;
    }
};

File: gfg/2024_8_august/python/08_sum_tree.py
# CODE

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class Solution:
    def is_sum_tree(self, node):
        if node is None:
            return True
            
        def solve(node):
            if node is None:
                return (True, 0)
            
            if (node.left is None) and (node.right is None):
                return (True, node.data)
            
            left_res = solve(node.left)
            right_res = solve(node.right)
            
            if (not left_res[0]) or (not right_res[0]):
                return (False, -1)
                
            return ((True, node.data + left_res[1] + right_res[1]) 
                     if node.data == (left_res[1] + right_res[1]) 
                     else (False, -1))
            
        return solve(node)[0]